peps/pep-0740.rst

```diff
@@ -205,40 +177,224 @@ changes to it:
+  ``provenance`` object in the ``file`` dictionary for that file.
+* ``provenance``, when present, **MUST** be a :ref:`provenance object <provenance-object>`.
+
+These changes require a version change to the JSON API:
+
+* The ``api-version`` must specify version 1.2 or later.
+
+.. _attestation-object:
+
+Attestation objects
+-------------------
+
+An attestation object is a JSON object with several required keys; applications
+or signers may include additional keys so long as all explicitly
+listed keys are provided. The required layout of an attestation
+object is provided as pseudocode below.
+
+.. code-block:: python
+
+  @dataclass
+  class Attestation:
+      version: Literal[1]
+      """
+      The attestation format's version, which is always 1.
+      """
+
+      verification_material: VerificationMaterial
+      """
+      Cryptographic materials used to verify `message_signature`.
+      """
+
+      message_signature: str
+      """
+      The attestation's signature, as `base64(raw-sig)`, where `raw-sig`
+      is the raw bytes of the signing operation.
+      """
+
+  @dataclass
+  class VerificationMaterial:
+      certificate: str
+      """
+      The signing certificate, as `base64(DER(cert))`.
+      """
+
+      transparency_entries: list[TransparencyLogEntry]
+      """
+      Zero or more transparency log entries for this attestation's signature
+      and certificate.
+      """
+
+  @dataclass
+  class TransparencyLogEntry:
+      log_index: int
+      """
+      The global index of the log entry, used when querying the log.
+      """
+
+      log_id: str
+      """
+      An opaque, unique identifier for the log.
+      """
+
+      entry_kind: str
+      """
+      The kind (type) of log entry.
+      """
+
+      entry_version: str
+      """
+      The version of the log entry's submitted format.
+      """
+
+      integrated_time: int
+      """
+      The UNIX timestamp from the log from when the entry was persisted.
+      """
+
+      inclusion_proof: InclusionProof
+      """
+      The actual inclusion proof the the log entry.
+      """
+
+
+  @dataclass
+  class InclusionProof:
+      log_index: int
+      """
+      The index of the entry in the tree it was written to.
+      """
+
+      root_hash: str
+      """
+      The digest stored at the root of the Merkle tree at the time of proof
+      generation.
+      """
+
+      tree_size: int
+      """
+      The size of the Merkle tree at the time of proof generation.
+      """
+
+      hashes: list[str]
+      """
+      A list of hashes required to complete the inclusion proof, sorted
+      in order from leaf to root. The leaf and root hashes are not themselves
+      included in this list; the root is supplied via `root_hash` and the client
+      must calculate the leaf hash.
+      """
+
+      checkpoint: str
+      """
+      The signed tree head's signature, at the time of proof generation.
+      """
```

Comment:
Flagging: I'm not sure if there's value in fully specifying these, or instead defining them behaviorally + leaving them as opaque objects in the PEP. Their structure is arguably an implementation detail, one with potential to change as logs (and log implementations) change.

Reply:
The only other field for completeness would be a list of cosigned checkpoints from witnesses. There is not a published spec yet for the format.

Reply:
Hmm, any thoughts on how I should represent that? `cosigned_checkpoints: list[object]`?

Reply:
`list[str]` should be good, it's going to look similar to the checkpoint format.